plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeWriter.java

```diff
@@ -264,39 +275,162 @@ public String toString()
                 .toString();
     }
 
-    private static final class TimestampTranslationBlockLoader
-            implements LazyBlockLoader
+    private static Optional<Function<Block, Block>> createTimestampCoercer(Type type)
+    {
+        if (type instanceof ArrayType arrayType) {
+            return Optional.of(new ArrayCoercer(arrayType));
+        }
+        if (type instanceof MapType mapType) {
+            return Optional.of(new MapCoercer(mapType));
+        }
+        if (type instanceof RowType rowType) {
+            return Optional.of(new RowCoercer(rowType));
+        }
+        if (type instanceof TimestampWithTimeZoneType) {
+            return Optional.of(new TimestampCoercer());
+        }
+        return Optional.empty();
+    }
+
+    private static class ArrayCoercer
+            implements Function<Block, Block>
     {
-        private Block originalBlock;
+        private final Optional<Function<Block, Block>> elementCoercer;
 
-        public TimestampTranslationBlockLoader(Block originalBlock)
+        public ArrayCoercer(ArrayType arrayType)
         {
-            this.originalBlock = requireNonNull(originalBlock, "originalBlock is null");
+            this.elementCoercer = createTimestampCoercer(arrayType.getElementType());
         }
 
         @Override
-        public Block load()
+        public Block apply(Block block)
+        {
+            if (elementCoercer.isEmpty()) {
+                return block;
+            }
+            ColumnarArray arrayBlock = toColumnarArray(block);
```

Comment:
Would it work if we coerce the child block (`onlyElement(block.getChildren())`) and keep offsets, nul blocks unchanged?

Reply:
(same for others)

Reply:
@findepi Do you mean the below code? Test passes with this code, but it requires changing protected `getOffsets` & `getValueIsNull` in `ArrayBlock` to public. 
```java
            ArrayBlock arrayBlock = (ArrayBlock) block;
            Block elementsBlock = elementCoercer.apply(getOnlyElement(block.getChildren()));
            return ArrayBlock.fromElementBlock(arrayBlock.getPositionCount(), Optional.ofNullable(arrayBlock.getValueIsNull()), arrayBlock.getOffsets(), elementsBlock);
```

Reply:
yes, i thought about something like that. 

cc @dain

Reply:
I'm not sure what you are doing, but the reason I added the `Columnar*` classes was to deal with all of the crazy edge cases.  For example, the example code in the comment, assums the block will be castable to an `ArrayBlock`, but what if it is a dictionary or RLE?  Additionally the element block might be a dictionary or RLE.  I recently simplified what is possible removing support for dictionaries in dictionaries and other combinations with RLE, but it is still complicated.... also you might need to deal with lazy blocks.

I'd stick with ColumnarArray unless you are very confident you have excellent test coverage.